avoid redundant c10::isCustomClassRegistered() checks

bhosmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42893 add CustomClassHolder inheritance check to boxing wrapper templates
* **#42852 avoid redundant c10::isCustomClassRegistered() checks**
* #42851 format make_boxed_from_unboxed_functor.h for readability

Differential Revision: [D23048381](https://our.internmc.facebook.com/intern/diff/D23048381)

Commit:
avoid redundant isCustomClassRegistered() checks

[ghstack-poisoned]

smessmer commented on aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:
```diff
@@ -59,13 +59,8 @@ namespace impl {
       guts::if_constexpr<guts::typelist::contains<supported_primitive_arg_types, T>::value>([] {
         /* everything is ok, this is a primitive type */
       }, /* else */ [] {
-        // TODO This is called for each operator call and potentially expensive.
```
We could still have a `std::is_base_of` check here just to make sure it actually inherits from the custom base class.

bhosmer replied:
Hmm looks like `is_base_of` reveals some use of classes that aren't in fact subclasses of `CustomClassHolder`. Since we're not actually reducing the test coverage if we omit the static tests from this PR, I'd rather land it as-is for the deduplication, and put the static checks (and whatever fixes are required) in a followup. Will add that one to the stack with additional info.

smessmer commented on aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:
```diff
@@ -171,9 +166,8 @@ namespace impl {
-        // TODO This is called for each operator call and potentially expensive.
-        // This check should be moved to operator registration time instead.
-        TORCH_CHECK(c10::isCustomClassRegistered<T>(), "Tried to use undefined class ", c10::util::get_fully_qualified_type_name<T>(), " as output");
+        /* otherwise this must be an instance of a valid custom class, since it can only
```
This comment is confusing because output types don't come from IValue, you don't have an IValue yet when calling into here. What you probably mean is something like "if this was invalid, it would fail later on when creating an IValue from it"

smessmer commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -1136,10 +1137,6 @@ IValue from_(T x, std::true_type) {
 }
 template <typename T>
```
let's enable_if this on an `is_base_of` check (or alternatively add a static_assert)

bhosmer replied:
See above re static check failing

smessmer commented on torch/custom_class.h:
```diff
@@ -261,11 +261,6 @@ class class_ {
 ///     IValue custom_class_iv = torch::make_custom_class<MyClass>(3, "foobarbaz");
 template <typename CurClass, typename... CtorArgs>
 c10::IValue make_custom_class(CtorArgs&&... args) {
```
same here

bhosmer replied:
See above re static check failing

Commit:
Update on "avoid redundant c10::isCustomClassRegistered() checks"


Differential Revision: [D23048381](https://our.internmc.facebook.com/intern/diff/D23048381)

[ghstack-poisoned]
